### PR TITLE
IB-16613 - Tensorflow based Canny edge detection module

### PR DESCRIPTION
**Pull request by arvindrajan92**

### CODE REVIEW

File: canny/canny.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+        # connected components
+        self.connected_components = connected_components.ConnectedComponents()
+
+    @tf.function(input_signature=[tf.TensorSpec(shape=[None, None, None], dtype=tf.float32)])
+    def __call__(self, images):
+        """Finds edges on the images employing Canny edge detection algorithm.
+
+        Args:
+            images: 3D tensor NHW (float32). Should be grayscale images.
```
Comment (FlatlinerDOA) [inline canny/canny.py:92]: @arvindrajan92 by NHW do you mean Nth Image, then Height, then Width?

File: canny/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+TensorFlow: 2.4.1
+Python: 3.8.10
+```
+The python packages and their versions used during the development of this module can be found in [`requirements.txt`](requirements.txt).
+
+#### Results
+
+```
+OpenCV: 0.0304 seconds for 100 reps of sudoku.png
+Tensorflow (Manual Thresholding): 0.9486 seconds for a batch of 100 sudoku.png
```
Comment (FlatlinerDOA) [inline canny/README.md:65]: @arvindrajan92 to clarify this was 1 batch of 100 images at once? And also that was GPU or CPU?

File: components/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+TensorFlow: 2.4.1
+Python: 3.8.10
+```
+The python packages and their versions used during the development of this module can be found in [`requirements.txt`](requirements.txt).
+
+#### Results
+
+```
+OpenCV: 0.0248 seconds for 1000 reps of eGaIy.jpg
+Tensorflow Addons: 0.9348 seconds for a batch of 1000 eGaIy.jpg
```
Comment (FlatlinerDOA) [inline components/README.md:47]: @arvindrajan92 this was 4 way only though right?

File: components/connected_components.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ 1. https://www.tensorflow.org/guide/gpu_performance_analysis
+ 2. https://github.com/NVIDIA/DeepLearningExamples/issues/57
+"""
+
+import tensorflow as tf
+import tensorflow_addons as tfa
+import numpy as np
+import os
+
+os.environ['TF_GPU_THREAD_MODE'] = 'gpu_private'
```
Comment (FlatlinerDOA) [inline components/connected_components.py:34]: @arvindrajan92 is this necessary for ConnectedComponents? Or is this left over from testing?

**Review by FlatlinerDOA — CHANGES_REQUESTED** [review]

File: canny/canny.py (hunk 1 above)
Comment (arvindrajan92) [reply canny/canny.py:92]: Yes, just the term on TensorFlow docs. C in NHWC would be Channel.

File: canny/README.md (hunk 2 above)
Comment (arvindrajan92) [reply canny/README.md:65]: Yes, this is 1 batch of N=100 images at once. 2 TF ops are not completely GPU-bound for hysteresis: tf.unique (used in hysteresis replacing) and [connected components](https://www.tensorflow.org/addons/api_docs/python/tfa/image/connected_components) used in the connected components module.

Given that it takes 10 ms per image and it scales linearly with N, I didn't spent too much time - cant really think of another way replacing connected components either for doing hysteresis.

File: components/README.md (hunk 3 above)
Comment (arvindrajan92) [reply components/README.md:47]: Yes, Tensorflow Addons is 4-way. Tensorflow (DNS) would be 8-way. Ran the 4-way for benchmark figures.

File: components/connected_components.py (hunk 4 above)
Comment (arvindrajan92) [reply components/connected_components.py:34]: There is some performance gain with setting the environment variable for TensorFlow: https://www.tensorflow.org/guide/gpu_performance_analysis. So I left it there.

**Pull request merged**